x/programs/rust/examples/single_nft/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+//! A basic NFT contract.
+//! The program serves as a non-fungible token with the ability to mint and burn.
+//! Only supports whole units with no decimal places.
+//!
+//! The NFT must support the common NFT metadata format.
+//! This includes the name, symbol, and URI of the NFT.
+use wasmlanche_sdk::{
+    memory::{Memory, Pointer},
+    program::Program,
+    public, state_keys,
+    types::Address,
+};
+
+/// The total supply of the NFT.
+/// # Safety
+/// In this program only one NFT of each type is intended to be minted.
+const TOTAL_SUPPLY: i64 = 1;
+
+/// The program storage keys.
+#[state_keys]
+enum StateKey {
+    /// The total supply of the token. Key prefix 0x0.
+    TotalSupply,
+    /// The name of the token. Key prefix 0x1.
+    Name,
+    /// The symbol of the token. Key prefix 0x2.
+    Symbol,
+    /// Metadata of the token. Key prefix 0x3.
+    Uri,
+    /// Balance of the NFT token by address. Key prefix 0x4(address).
+    Balance(Address),
+    /// Counter -- used to keep track of total NFTs minted. Key prefix 0x5.
+    Counter,
+    /// Owner -- used to keep track of the owner. Key prefix 0x6.
+    Owner,
+}
+
+/// Initializes the NFT with all required metadata.
+/// This includes the name, symbol, image URI, owner, and total supply.
+/// Returns true if the initialization was successful.
+#[public]
+pub fn init(
+    program: Program,
+    name_ptr: i64,
+    nft_name_length: i64,
```

Comment:
Can we not use the `string` type for this? Feel like we should use that macro (don't want people doing this memory parsing manually)

Reply:
We can't use a `String`, wasm only works on ints and floats.
We could try for a macro but I'm not sure how to implement this. I can check. 
 cc @hexfusion

Reply:
Can you outline what a macro would do in this case to solve this problem?